internal/controller/nicdevice_controller.go

```diff
@@ -0,0 +1,144 @@
+/*
+2024 NVIDIA CORPORATION & AFFILIATES
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controller
+
+import (
+	"context"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/Mellanox/nic-configuration-operator/api/v1alpha1"
+	"github.com/Mellanox/nic-configuration-operator/pkg/host"
+)
+
+// NicDeviceReconciler reconciles a NicDevice object
+type NicDeviceReconciler struct {
+	client.Client
+	Scheme *runtime.Scheme
+
+	nodeName    string
+	namespace   string
+	hostManager host.HostManager
+	hostUtils   host.HostUtils
+}
+
+//+kubebuilder:rbac:groups=configuration.net.nvidia.com,resources=nicdevices,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=configuration.net.nvidia.com,resources=nicdevices/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=configuration.net.nvidia.com,resources=nicdevices/finalizers,verbs=update
+
+// Reconcile reconciles the NicDevice object and applies the configuration defined in spec
+func (r *NicDeviceReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	device := &v1alpha1.NicDevice{}
+
+	err := r.Get(ctx, req.NamespacedName, device)
+	if err != nil {
+		return reconcile.Result{}, client.IgnoreNotFound(err)
+	}
+
+	if device.Spec.Configuration == nil {
+		return reconcile.Result{}, nil
+	}
+
+	err = r.UpdateDeviceStatus(ctx, device)
```

Comment:
Currently, If the spec is empty the status will not be updated with the "spec not set" condition. Is this intentional?